# origin: roehrich (BitKeeper)
# cset: 1.1449.17.22 (2.4) key=41240cebupWd4HFyNZ6iso1kB1f4IA
# inclusion: upstream
# description: [XFS] Fix lock leak in xfs_free_file_space
# revision date: Fri, 26 Nov 2004 14:59:38 +0900
#
# S rset: ChangeSet|1.1449.17.21..1.1449.17.22
# I rset: fs/xfs/xfs_vnodeops.c|1.178..1.179
#
# This is a BitKeeper generated diff -Nru style patch.
#
# ChangeSet
#   2004/08/19 12:14:03+10:00 roehrich@sgi.com 
#   [XFS] Fix lock leak in xfs_free_file_space
#   
#   SGI Modid: xfs-linux:xfs-kern:176905a
#   Signed-off-by: Nathan Scott <nathans@sgi.com>
# 
# fs/xfs/xfs_vnodeops.c
#   2004/08/19 12:13:57+10:00 roehrich@sgi.com +6 -2
#   [XFS] Fix lock leak in xfs_free_file_space
# 
#
===== fs/xfs/xfs_vnodeops.c 1.178 vs 1.179 =====
--- 1.178/fs/xfs/xfs_vnodeops.c	2004-06-18 09:59:16 +09:00
+++ 1.179/fs/xfs/xfs_vnodeops.c	2004-08-19 11:13:57 +09:00
@@ -4337,8 +4337,10 @@
 		nimap = 1;
 		error = xfs_bmapi(NULL, ip, startoffset_fsb, 1, 0, NULL, 0,
 			&imap, &nimap, NULL);
-		if (error)
+		if (error) {
+			xfs_iunlock(ip, XFS_IOLOCK_EXCL);
 			return error;
+		}
 		ASSERT(nimap == 0 || nimap == 1);
 		if (nimap && imap.br_startblock != HOLESTARTBLOCK) {
 			xfs_daddr_t	block;
@@ -4352,8 +4354,10 @@
 		nimap = 1;
 		error = xfs_bmapi(NULL, ip, endoffset_fsb - 1, 1, 0, NULL, 0,
 			&imap, &nimap, NULL);
-		if (error)
+		if (error) {
+			xfs_iunlock(ip, XFS_IOLOCK_EXCL);
 			return error;
+		}
 		ASSERT(nimap == 0 || nimap == 1);
 		if (nimap && imap.br_startblock != HOLESTARTBLOCK) {
 			ASSERT(imap.br_startblock != DELAYSTARTBLOCK);
